### PR TITLE
Mark mono_class_vtable and mono_class_from_typeref external only.

### PR DESCRIPTION
**Pull request by lambdageek**
- Runtime should use `mono_class_vtable_checked` and `mono_class_from_typeref_checked`, respectively.
- Rename `mono_class_vtable_full` to `mono_class_vtable_checked`
- Pass a `MonoError*` argument to
  - `mono_marshal_get_remoting_invoke`
  - `mono_marshal_get_xappdomain_invoke`
  - `mono_marshal_get_remoting_invoke_for_target`
  - `mono_marshal_get_remoting_invoke_with_check`

### CODE REVIEW

**Comment by lambdageek** [comment]
@lewurm could you take a look at the error handling in `interp/` to make sure it makes sense?

@vargaz Same request for the JIT - I mostly routed the errors to `&cfg->error`, which seemed like the right thing.

File: mono/mini/interp/transform.c [hunk 1]
```diff
@@ -4450,8 +4452,13 @@ mono_interp_transform_method (InterpMethod *imethod, ThreadContext *context, Int
 					return mono_error_convert_to_exception (&error);
 				}
 				mono_class_init (m->klass);
-				if (!mono_class_is_interface (m->klass))
-					mono_class_vtable (domain, m->klass);
+				if (!mono_class_is_interface (m->klass)) {
+					mono_class_vtable_full (domain, m->klass, &error);
+					if (!is_ok (&error)) {
+						g_free (is_bb_start);
+						return mono_error_convert_to_exception (&error);
```
Comment (lewurm) [inline mono/mini/interp/transform.c]: not exactly obvious, but it should also call `mono_metadata_free_mh (header);`.  On the other hand, how often are we going to hit this case?

File: mono/metadata/object-internals.h [hunk 2]
```diff
@@ -85,15 +85,17 @@
...
+	if (is_ok ((error)))						\
+		__arr = mono_array_new_specific_checked (__vtable, (size), (error)); \
+	__arr; })
 
 #else
 
 #define mono_class_get_field_from_name_cached(klass,name) mono_class_get_field_from_name ((klass), (name))
 #define mono_array_class_get_cached(eclass,rank) mono_array_class_get ((eclass), (rank))
-#define mono_array_new_cached(domain, eclass, size, error) mono_array_new_specific_checked (mono_class_vtable ((domain), mono_array_class_get_cached ((eclass), 1)), (size), (error))
+#define mono_array_new_cached(domain, eclass, size, error) mono_array_new_checked ((domain), (eclass), (size), (error))
```
Comment (lewurm) [inline mono/metadata/object-internals.h]: this change looks a bit odd; is it intended?

Comment (lambdageek) [reply mono/metadata/object-internals.h]: Yes. it's intentional.  note that in this block `mono_array_class_get_cached` is defined as `mono_array_class_get`.  With that, the old code amounts to exactly the same thing as the implementation of `mono_array_new_checked`.

**Pull request merged**